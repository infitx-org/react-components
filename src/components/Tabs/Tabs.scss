@import "../scss/vars/colors.scss";
@import "../scss/base.scss";

.rc-tabs {
  outline: none;
  @include rc-component;
}

.rc-tabs--flexible {
  display: flex;
  flex-direction: column;
  flex-grow: 1;
  min-height: 0;
}

.rc-tabs__tabs {
  display: flex;
  flex-direction: row;
  flex: 0 0 auto;
}

.rc-tabs__tab {
  flex: 0 1 auto;
  padding: 0px 10px;
  height: 32px;
  font-size: 14px;
  line-height: 32px;
  text-align: center;
  color: $color__dark-gray-3;
  border-bottom: 3px solid $color__gray-2;
  transition: all 0.1s ease-out;
  cursor: pointer;

  &.rc-tabs__tab--flexible {
    flex-grow: 1;
  }

  &:hover:not(.rc-tabs__tab--disabled):not(.rc-tabs__tab--selected) {
    border-bottom-color: $color__gray-4;
  }
  &.rc-tabs__tab--selected {
    border-bottom-color: $color__main-primary;
    color: $color__main-primary;
  }
  &.rc-tabs__tab--focused {
    border-bottom-color: $color__main-primary--active;
    color: $color__main-primary--active;
  }
  &.rc-tabs__tab--disabled {
    color: $color__main-disabled;
    cursor: default;
  }
}

.rc-tabs__tab-panel {
  &.rc-tabs__tab-panel--flexible {
    display: flex;
    flex-grow: 1;
    flex-direction: column;
    min-height: 0;
  }
}
