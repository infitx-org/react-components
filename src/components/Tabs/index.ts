export { Tab, Tabs, TabPanel } from "./Tabs";
export type { TabPanelProps, TabProps, TabsProps } from "./Tabs";
