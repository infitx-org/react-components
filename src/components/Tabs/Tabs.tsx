import classnames from "classnames";
import React, {
  PureComponent,
  KeyboardEvent,
  MouseEvent,
  FocusEvent,
} from "react";
import "./Tabs.scss";

/* eslint-disable jsx-a11y/no-noninteractive-tabindex */

interface TabProps {
  children: React.ReactNode;
  selected?: boolean;
  focused?: boolean;
  disabled?: boolean;
  hidden?: boolean;
  flex?: boolean;
  style?: React.CSSProperties;
  className?: string;
  onSelect?: (e: MouseEvent<HTMLDivElement>) => void;
}

type TabWithRefProps = TabProps & {
  ref: React.Ref<HTMLDivElement>;
};

function Tab(
  {
    selected,
    focused,
    children,
    disabled,
    hidden,
    flex,
    style,
    className,
    onSelect,
  }: TabProps,
  ref: React.Ref<HTMLDivElement>
) {
  if (hidden) {
    return null;
  }
  const tabClassName = classnames([
    "el",
    "el-tabs__tab",
    focused && "el-tabs__tab--focused",
    selected && "el-tabs__tab--selected",
    disabled && "el-tabs__tab--disabled",
    flex && "el-tabs__tab--flexible",
    className,
  ]);

  return (
    <div
      ref={ref}
      onClick={onSelect}
      className={tabClassName}
      style={style}
      role="presentation"
    >
      {children}
    </div>
  );
}

const TabWithRef = React.forwardRef<HTMLDivElement, TabProps>(Tab);

interface TabPanelProps {
  style?: React.CSSProperties;
  className?: string;
  flex?: boolean;
  children: React.ReactNode;
}

function TabPanel({ children, flex, style, className }: TabPanelProps) {
  const classNames = classnames([
    "el-tabs__tab-panel",
    flex && "el-tabs__tab-panel--flexible",
    className,
  ]);
  return (
    <div className={classNames} style={style}>
      {children}
    </div>
  );
}

type TabElement = React.ReactElement<TabWithRefProps>;
type TabPanelElement = React.ReactElement<TabPanelProps>;

interface TabsProps {
  style?: React.CSSProperties;
  id?: string;
  className?: string;
  selected?: number;
  disabled?: boolean;
  flex?: boolean;
  children: TabElement[] | TabPanelElement[];
  onSelect?: (evt: MouseEvent, index: number) => void;
  onFocus?: (evt: FocusEvent) => void;
  onBlur?: (evt: FocusEvent) => void;
}

interface TabsState {
  selected: number;
  hasFocus?: boolean;
}

class Tabs extends PureComponent<TabsProps, TabsState> {
  static isTabPanel(child: React.ReactNode): boolean {
    return (child as React.ReactElement).type === TabPanel;
  }

  static isTab(child: React.ReactNode): boolean {
    return (child as React.ReactElement).type === TabWithRef;
  }

  private tabRefs = React.createRef<HTMLDivElement[]>([]);

  constructor(props: TabsProps) {
    super(props);

    this.state = {
      selected: this.getSelectableIndex(this.props.selected),
      hasFocus: false,
    };

    this.getComponents = this.getComponents.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.selectTab = this.selectTab.bind(this);

    this.onSelect = this.onSelect.bind(this);
    this.onFocus = this.onFocus.bind(this);
    this.onBlur = this.onBlur.bind(this);
  }

  handleKeyDown(e: KeyboardEvent): void {
    const { keyCode } = e.nativeEvent;
    if (keyCode === 37) {
      this.selectTab(this.state.selected - 1, false);
      return;
    }
    if (keyCode === 39) {
      this.selectTab(this.state.selected + 1, true);
    }
  }

  onFocus(e: FocusEvent) {
    this.setState(
      {
        hasFocus: true,
      },
      () => this.props.onFocus?.(e)
    );
  }

  onBlur(e: FocusEvent) {
    this.setState(
      {
        hasFocus: false,
      },
      () => this.props.onBlur?.(e)
    );
  }

  onSelect(evt: MouseEvent, index: number): void {
    this.setState({ selected: index, hasFocus: true }, () => {
      this.props.onSelect?.(evt, index);
    });
  }

  getComponents(): [TabElement[], TabPanelElement[]] {
    const components = React.Children.toArray(this.props.children);
    const tabs = components.filter(Tabs.isTab) as TabElement[];
    const tabPanels = components.filter(Tabs.isTabPanel) as TabPanelElement[];
    return [tabs, tabPanels];
  }

  getSelectableIndex(index: number = 0, goNext: boolean = true): number {
    function arrayRotate<T>(src: T[], count: number, reverse: boolean): T[] {
      const arr = [...src];
      for (let i = 0; i < count; i += 1) {
        if (reverse) arr.unshift(arr.pop() as T);
        else arr.push(arr.shift() as T);
      }
      return arr;
    }
    const [tabs] = this.getComponents();
    const r = arrayRotate(tabs, Math.abs(index), !goNext);
    const inOrder = (goNext ? r : r.reverse()).find((t) => !t.props.disabled);
    return inOrder ? tabs.indexOf(inOrder) : -1;
  }

  selectTab(index: number, goNext: boolean): void {
    const toClick = this.getSelectableIndex(index, goNext);
    this.tabRefs.current?.[toClick].click();
  }

  render() {
    const { selected } = this.state;
    const { id, style, className, disabled } = this.props;
    const [rawTabs, rowPanels] = this.getComponents();

    const tabs = rawTabs.map((tab, index) => {
      const isDisabled = tab.props.disabled || disabled;
      const isSelected = selected === index;
      const isFocused = isSelected && this.state.hasFocus;
      const onSelect = !isDisabled
        ? (evt: MouseEvent) => this.onSelect(evt, index)
        : undefined;
      return React.cloneElement(tab, {
        ...tab.props,
        key: index.toString(),
        disabled: isDisabled,
        selected: isSelected,
        focused: isFocused,
        onSelect,
        ref: (node: HTMLDivElement) => {
          this.tabRefs.current = [...(this.tabRefs.current || []), node];
        },
      });
    });

    const panel = rowPanels[selected] || null;
    const tabsClassNames = classnames(["element", "el-tabs", className]);

    return (
      <div
        id={id}
        style={style}
        className={tabsClassNames}
        tabIndex={0}
        role="presentation"
        onKeyDown={this.handleKeyDown}
        onFocus={this.onFocus}
        onBlur={this.onBlur}
      >
        <div className="el-tabs__tabs">{tabs}</div>
        {panel}
      </div>
    );
  }
}

export { TabWithRef as Tab, Tabs, TabPanel };
