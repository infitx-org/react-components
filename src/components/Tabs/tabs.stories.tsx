/* eslint no-console: "off" */
import React from "react";

import { Tab, TabPanel, Tabs } from "./Tabs";

export default {
  title: "Components/Tabs",
  component: Tabs,
};

export const Default = () => (
  <Tabs>
    <Tab>Tab 1</Tab>
    <Tab>Tab 2</Tab>
    <Tab>Tab 3</Tab>
    <TabPanel> Tab Content 1 </TabPanel>
    <TabPanel> Tab Content 2 </TabPanel>
    <TabPanel> Tab Content 3 </TabPanel>
  </Tabs>
);

export const DisabledTab = () => (
  <Tabs selected={0} onSelect={console.log}>
    <Tab>Tab 1</Tab>
    <Tab>Tab 2</Tab>
    <Tab disabled>Tab 3</Tab>
    <Tab>Tab 4</Tab>
    <Tab>Tab 5</Tab>
    <TabPanel> Tab Content 1 </TabPanel>
    <TabPanel> Tab Content 2 </TabPanel>
    <TabPanel> Tab Content 3 </TabPanel>
    <TabPanel> Tab Content 4 </TabPanel>
    <TabPanel> Tab Content 5 </TabPanel>
  </Tabs>
);

export const StyledTab = () => (
  <Tabs>
    <Tab>Tab 1</Tab>
    <Tab>Tab 2</Tab>
    <Tab style={{ width: "300px", background: "#def", fontWeight: "bold" }}>
      Tab 3 (styled)
    </Tab>
    <TabPanel> Tab Content 1 </TabPanel>
    <TabPanel> Tab Content 2 </TabPanel>
    <TabPanel> Tab Content 3 </TabPanel>
  </Tabs>
);

export const FlexTab = () => (
  <Tabs>
    <Tab>Tab 1</Tab>
    <Tab>Tab 2</Tab>
    <Tab flex>Tab 3 (flex)</Tab>
    <TabPanel> Tab Content 1 </TabPanel>
    <TabPanel> Tab Content 2 </TabPanel>
    <TabPanel> Tab Content 3 </TabPanel>
  </Tabs>
);

export const StyledTabPanel = () => (
  <Tabs>
    <Tab>Tab 1</Tab>
    <Tab>Tab 2</Tab>
    <Tab>Tab 3 (panel styled)</Tab>
    <TabPanel
      style={{ width: "300px", background: "#def", fontWeight: "bold" }}
    >
      {" "}
      Tab Content 1{" "}
    </TabPanel>
    <TabPanel> Tab Content 2 </TabPanel>
    <TabPanel> Tab Content 3 </TabPanel>
  </Tabs>
);

// const flexTabPanelStyle = {
//   background: "#def",
//   alignItems: "center",
//   justifyContent: "center",
// };

// export const Flexible = () => (
//   <div style={{ height: "200px", background: "#f8f8f8" }}>
//     <Tabs flex>
//       <TabList>
//         <Tab>Tab 1</Tab>
//         <Tab>Tab 2</Tab>
//         <Tab>Tab 3</Tab>
//       </TabList>
//       <TabPanels>
//         <TabPanel style={flexTabPanelStyle}> Tab Content 1 </TabPanel>
//         <TabPanel style={flexTabPanelStyle}> Tab Content 2 </TabPanel>
//         <TabPanel style={flexTabPanelStyle}> Tab Content 3 </TabPanel>
//       </TabPanels>
//     </Tabs>
//   </div>
// );

// export const SingleTabFlexible = () => (
//   <div style={{ height: "200px", background: "#f8f8f8" }}>
//     <Tabs>
//       <TabList>
//         <Tab>Tab 1</Tab>
//         <Tab>Tab 2</Tab>
//         <Tab>Tab 3</Tab>
//       </TabList>
//       <TabPanels>
//         <TabPanel flex style={flexTabPanelStyle}>
//           {" "}
//           Flexible Tab Content 1{" "}
//         </TabPanel>
//         <TabPanel style={flexTabPanelStyle}> Tab Content 2 </TabPanel>
//         <TabPanel style={flexTabPanelStyle}> Tab Content 3 </TabPanel>
//       </TabPanels>
//     </Tabs>
//   </div>
// );
