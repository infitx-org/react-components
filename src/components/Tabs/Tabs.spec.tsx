import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import { Tab, TabPanel, Tabs } from "./Tabs";

const testTabs = (
  <Tabs id="test-tabs" flex>
    <Tab>Tab1</Tab>
    <Tab>Tab2</Tab>
    <Tab disabled>Tab3</Tab>
    <TabPanel>TabPanel1</TabPanel>
    <TabPanel>TabPanel2</TabPanel>
    <TabPanel>TabPanel3</TabPanel>
  </Tabs>
);

describe("tests the tabs", () => {
  it("renders the tabs", () => {
    const { container } = render(testTabs);
    expect(container.querySelector(".el-tabs")).toBeTruthy();
  });

  it("renderd the prop id", () => {
    const { container } = render(testTabs);
    expect(container.querySelector("#test-tabs")).toBeTruthy();
  });

  it("renders all the tab items", () => {
    const { container } = render(testTabs);
    expect(container.querySelectorAll(".el-tabs__tab")).toHaveLength(3);
  });

  it("renders only one tab panel", () => {
    const { container } = render(testTabs);
    expect(container.querySelectorAll(".el-tabs__tab-panel")).toHaveLength(1);
  });

  it("renders the first tabpanel content", () => {
    const { container } = render(testTabs);
    expect(screen.getByText("TabPanel1")).toBeInTheDocument();
  });

  it("renders the second tabpanel content when second tab is clicked", () => {
    const { container } = render(testTabs);
    const [, secondTab] = container.querySelectorAll(".el-tabs__tab");
    userEvent.click(secondTab);
    expect(screen.getByText("TabPanel2")).toBeInTheDocument();
  });

  it("does not renders the tabpanel of a disabled tab when clicked", () => {
    const { container } = render(testTabs);
    const [, , thirdTab] = container.querySelectorAll(".el-tabs__tab");
    userEvent.click(thirdTab);
    expect(screen.queryByText("TabPanel3")).not.toBeInTheDocument();
  });

  it("triggers onSelect when clicking an unselected tab", () => {
    const mockEvent = jest.fn();
    const { container } = render(
      <Tabs onSelect={mockEvent}>
        <Tab>Tab1</Tab>
        <Tab>Tab2</Tab>
      </Tabs>
    );

    const [, secondTab] = container.querySelectorAll(".el-tabs__tab");
    userEvent.click(secondTab);
    expect(mockEvent).toHaveBeenCalled();
    // extract the tab this way
    const [clickEvent] = mockEvent.mock.calls[0];
    expect(clickEvent.type).toEqual("click");
    expect(clickEvent.target).toEqual(secondTab);
  });

  it("does not triggers onSelect when clicking an disabled tab", () => {
    const mockEvent = jest.fn();
    const { container } = render(
      <Tabs onSelect={mockEvent}>
        <Tab>Tab1</Tab>
        <Tab disabled>Tab2</Tab>
      </Tabs>
    );

    const [, secondTab] = container.querySelectorAll(".el-tabs__tab");
    userEvent.click(secondTab);
    expect(mockEvent).not.toHaveBeenCalled();
  });

  it("does not triggers onSelect when clicking the selected tab", () => {
    const mockEvent = jest.fn();
    const { container } = render(
      <Tabs onSelect={mockEvent}>
        <Tab>Tab1</Tab>
        <Tab>Tab2</Tab>
      </Tabs>
    );

    const [firstTab] = container.querySelectorAll(".el-tabs__tab");
    userEvent.click(firstTab);
    expect(mockEvent).not.toHaveBeenCalled();
  });
});

// Snapshot

it("renders the tabs correctly when multiple props are set", () => {
  const { container } = render(testTabs);
  expect(container).toMatchSnapshot();
});
