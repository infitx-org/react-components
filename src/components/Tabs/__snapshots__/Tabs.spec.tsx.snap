// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the tabs correctly when multiple props are set 1`] = `
<div>
  <div
    class="element rc-tabs"
    id="test-tabs"
    role="presentation"
    tabindex="0"
  >
    <div
      class="rc-tabs__tabs"
    >
      <div
        class="el rc-tabs__tab rc-tabs__tab--selected"
        role="presentation"
      >
        Tab1
      </div>
      <div
        class="el rc-tabs__tab"
        role="presentation"
      >
        Tab2
      </div>
      <div
        class="el rc-tabs__tab rc-tabs__tab--disabled"
        role="presentation"
      >
        Tab3
      </div>
    </div>
    <div
      class="el rc-tabs__tab-panel"
    >
      TabPanel1
    </div>
  </div>
</div>
`;
